File: elixir/cell.ex
defmodule Cell do
  defstruct [:x, :y, :alive, :neighbours]

  def new(x, y, alive \\ false) do
    %Cell{
      x: x,
      y: y,
      alive: alive
    }
  end

  def to_char(cell) do
    if cell.alive, do: "o", else: " "
  end

  def alive_neighbours(cell, world) do
    neighbours = Map.take(world.cells, cell.neighbours)

    # The following is slower
    # Enum.count(neighbours, fn {_key, neighbour} ->
    #   neighbour.alive
    # end)

    # The following is the fastest
    for {_key, neighbour} <- neighbours, reduce: 0 do
      count ->
        if neighbour.alive do
          count + 1
        else
          count
        end
    end
  end
end

File: erlang/cell.erl
-module(cell).
-include("cell.hrl").

-export([new/2, new/3, to_char/1, alive_neighbours/2]).
-export_type([cell/0]).

-spec new(integer(), integer()) -> cell().
new(X, Y) ->
  new(X, Y, false).

-spec new(integer(), integer(), boolean()) -> cell().
new(X, Y, Alive) ->
  #cell{
    x = X,
    y = Y,
    alive = Alive
  }.

-spec to_char(cell()) -> char().
to_char(#cell{alive = true}) -> $o;
to_char(#cell{alive = false}) -> $\s.

-spec alive_neighbours(cell(), world:world()) -> integer().
alive_neighbours(Cell, World) ->
  Neighbours = maps:with(Cell#cell.neighbours, world:cells(World)),

  %% The following is slower
  %% maps:fold(
  %%   fun(_Key, Neighbour, Count) ->
  %%     case Neighbour#cell.alive of
  %%       true -> Count + 1;
  %%       false -> Count
  %%     end
  %%   end,
  %%   0,
  %%   Neighbours
  %% ).

  %% The following is the fastest
  length([Neighbour || Neighbour <- maps:values(Neighbours), Neighbour#cell.alive]).
